File: Курс 1/Семестр 1/DM/lab1/F.py
def insert(mas, vec):
    s = ""
    c = 0
    for i in range(len(mas)):
        if mas[i] >= 0:
            s += str(mas[i])
        else:
            s += vec[c]
            c += 1
    return s

n, k = map(int, input().split())
M = []
for i in range(k):
    A = list(map(int, input().split()))
    cnt_of_absents = 0
    for j in range(n):
        if A[j] == 1:
            A[j] = 0
        elif A[j] == 0:
            A[j] = 1
        elif A[j] == -1:
            cnt_of_absents += 1
    M.append([A, cnt_of_absents])

C = [0] * 2**n
for i in range(k):
    for j in range(2**M[i][1]):
        if M[i][1] != 0:
            vec_j = bin(j)[2:]
            vec_j = "0" * (M[i][1] - len(vec_j)) + vec_j
            num = int(insert(M[i][0], vec_j), 2)
            C[num] = 1
        else:
            num = int(insert(M[i][0], ""), 2)
            C[num] = 1

if 0 in C:
    print("NO")
else:
    print("YES")

File: Курс 1/Семестр 1/DM/lab1/F.java
import java.util.Scanner;

public class F {

    public static String insert(int[] mas, String vec) {
        StringBuilder s = new StringBuilder();
        int c = 0;
        for (int i = 0; i < mas.length; i++) {
            if (mas[i] >= 0) {
                s.append(mas[i]);
            } else {
                s.append(vec.charAt(c));
                c++;
            }
        }
        return s.toString();
    }

    public static String toBinary(int num, int w) {
        int d = num;
        StringBuilder ans = new StringBuilder();
        while (d != 0) {
            ans.append((d % 2));
            d = d / 2;
        }
        ans.append("0".repeat(w - ans.length()));
        return ans.reverse().toString();
    }

    public static int toDecimal(String num) {
        int ans = 0;
        for (int i = 0; i < num.length(); i++) {
            ans += Integer.parseInt(Character.toString(num.charAt(i))) * (1 << (num.length() - i - 1));
        }
        return ans;
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt(), k = sc.nextInt();
        int[][] M = new int[k][n];
        int[] cnt_of_absents = new int[k];
        int temp = 0;

        for (int i = 0; i < k; i++) {
            cnt_of_absents[i] = 0;
            for (int j = 0; j < n; j++) {
                temp = sc.nextInt();
                if (temp == 1) {
                    M[i][j] = 0;
                } else if (temp == 0) {
                    M[i][j] = 1;
                } else if (temp == -1) {
                    M[i][j] = -1;
                    cnt_of_absents[i]++;
                }
            }
        }

        int[] C = new int[1 << n];
        int t = 0;
        int num = 0;
        StringBuilder vec_j = new StringBuilder();
        for (int i = 0; i < k; i++) {
            for (int j = 0; j < (1 << cnt_of_absents[i]); j++) {
                if (cnt_of_absents[i] != 0) {
                    vec_j.setLength(0);
                    vec_j.append(toBinary(j, cnt_of_absents[i]));
                    num = toDecimal(insert(M[i], vec_j.toString()));
                    if (C[num] == 0) {
                        C[num] = 1;
                        t++;
                    }
                } else {
                    num = toDecimal(insert(M[i], ""));
                    if (C[num] == 0) {
                        C[num] = 1;
                        t++;
                    }
                }
            }


        }
        if (t == (1 << n)) {
            System.out.println("YES");
        } else {
            System.out.println("NO");
        }
    }
}
